File: server/api-client/market.js
// 统一行情接口：优先用 EastMoney（字段全），失败后自动切换到腾讯/新浪
const eastmoney = require("./eastmoney");
const { stocks: providers } = require("../providers");
const { DEFAULT_STOCK } = require("../utils/constant");

const auto = providers.auto;

// 将 stock-api 的 7 字段格式转为统一 14 字段格式
function normalizeStock(s) {
  return {
    code: s.code || "",
    name: s.name || "---",
    price: s.now ?? null,
    changePercent: s.percent != null ? s.percent * 100 : null, // 0.01 → 1%
    open: null,
    high: s.high ?? null,
    low: s.low ?? null,
    yesterday: s.yesterday ?? null,
    volume: null,
    amount: null,
    amplitude: null,
    turnoverRate: null,
    totalMarketCap: null,
    source: s.source || "auto",
  };
}

// 统一搜索
async function searchStocks(keyword) {
  try {
    // 优先用 eastmoney 搜索（字段全）
    const em = await eastmoney.searchStocks(keyword);
    if (em.length > 0) return em;
  } catch {}
  // 降级：用 auto provider 搜索（多源 failover）
  try {
    const results = await auto.searchStocks(keyword);
    return results.filter((s) => s.name && s.name !== DEFAULT_STOCK.name).map(normalizeStock);
  } catch {
    return [];
  }
}

// 统一批量查询
async function getStocks(codes) {
  if (!codes.length) return [];
  // 先试试 eastmoney
  try {
    const em = await eastmoney.getStocks(codes);
    if (em.length > 0) return em;
  } catch {}
  // 降级：用 auto provider
  try {
    const results = await auto.getStocks(codes);
    return results
      .filter((s) => s.name && s.name !== DEFAULT_STOCK.name)
      .map(normalizeStock);
  } catch {
    return codes.map((c) => ({ code: c, name: "---" }));
  }
}

// 统一单只查询
async function getStock(code) {
  const stocks = await getStocks([code]);
  return stocks[0] || null;
}

module.exports = { searchStocks, getStocks, getStock };

File: server-py/app/services/market.py
"""行情查询服务 —— 协调 provider 与数据库"""

from sqlalchemy.orm import Session

from app.models import Stock, Watchlist
from clients import auto


async def search_stocks(keyword: str) -> list[dict]:
    return await auto.search_stocks(keyword)


async def get_stock(code: str, db: Session) -> dict | None:
    """查单只股票：优先 DB，DB 没有则调 API 并存入 DB"""
    row = (
        db.query(Stock)
        .filter(Stock.code == code)
        .order_by(Stock.created_at.desc())
        .first()
    )
    if row:
        return _row_to_dict(row)

    stock = await auto.get_stock(code)
    if stock and stock.get("name") and stock["name"] != "---":
        stock["code"] = code
        _insert_stock(db, stock)
    return stock


async def get_stocks(codes: list[str]) -> list[dict]:
    return await auto.get_stocks(codes)


def _insert_stock(db: Session, stock: dict):
    row = Stock(
        code=stock.get("code"),
        name=stock.get("name"),
        price=stock.get("price"),
        changePercent=stock.get("changePercent"),
        open=stock.get("open"),
        high=stock.get("high"),
        low=stock.get("low"),
        yesterday=stock.get("yesterday"),
        volume=stock.get("volume"),
        amount=stock.get("amount"),
        amplitude=stock.get("amplitude"),
        turnoverRate=stock.get("turnoverRate"),
        totalMarketCap=stock.get("totalMarketCap"),
        source=stock.get("source", "eastmoney"),
    )
    db.add(row)
    db.commit()


def insert_stock(db: Session, stock: dict):
    _insert_stock(db, stock)


def _row_to_dict(row: Stock) -> dict:
    return {
        "code": row.code,
        "name": row.name,
        "price": _d(row.price),
        "changePercent": _d(row.changePercent),
        "open": _d(row.open),
        "high": _d(row.high),
        "low": _d(row.low),
        "yesterday": _d(row.yesterday),
        "volume": _d(row.volume),
        "amount": _d(row.amount),
        "amplitude": _d(row.amplitude),
        "turnoverRate": _d(row.turnoverRate),
        "totalMarketCap": _d(row.totalMarketCap),
        "source": row.source,
        "created_at": row.created_at.isoformat() if row.created_at else None,
    }


def _d(val) -> float | None:
    return float(val) if val is not None else None
